### ISSUE TITLE
[Feature Request] ability to work as a DNS resolver

### ISSUE THREAD

**Issue by lordraiden** [AdguardTeam/AdGuardHome#790]
Right now Adguard just fwd the DNS requests to "public" DNS of our choice.
It would be nice to have the ability to call directly to the root servers acting as a DNS resolver.
I think it will improve privacy a lot but maybe it wont be as fast, but is a tradeoff

**Comment by ameshkov** [comment]
We're looking into it right now actually.

Basically, we have two options:
1. Use a resolver implementation written in pure Go
2. Use `libunbound` + CGO bindings

`libunbound` is a stable and reputable software, but the drawback is that it complicates things, and it'd be hard to cross-compile it for every OS.

**Comment by lordraiden** [comment]
Maybe you already know this list, but it can give you some idea of what solutions are out there.
https://doh.defaultroutes.de/implementations.html

**Comment by ibksturm** [comment]
hi there

i had similar thinks when i start with ibksturm.synology.me....
a) being a resolver to root
b) being a forwarder

i choose a combination. opennic i running with nsd a copy of root and iana forward

my reasons are:
- opennic local copy: cause being a duplicate of the hole alt. dns system and it isn‘t speed „important“
- iana forwards because
 - speed, forwarding is much faster than resolving (i use resolving only as fallback)
 - privacy reason, all forwards are over tls 853. unbound change in the actualy modified config over 6 dns server, where the fastest ist weighted to 40%.
 i also wrote months ago to the iana guys a mail where i could get encryted root server. the answer was that all root servers are for reason unencrypted and thats at the moment not a option in their roadmap
- les data traffic. if you‘ve got over 1.5 mio request / 24h you will think about.... keynote: myisp had also think about and installed on street side a new fiberswitch station 👍 it seams that i crushed hole neighbourhood :) 

back to original question

i think it would be better for future of project if you wrote it complettly newly on Go.

+ you „ know“ better what your code do
+ you could splitt it into a side project as dnsproxy
+ as you wrote its easyier to compile

- there could be some handmade security bugs where unbound guys are well known


maybe if i thinking loud, is there any reference on unbound websides? or maybe theguys from unbound had allready done a Go project. Wrote them a mail  wouldn costs a lot... just trying to help you guys

greets

**Comment by ameshkov** [comment]
> there could be some handmade security bugs where unbound guys are well known

Well, there is a much greater chance that there will be bugs in our implementation :)

**Comment by ibksturm** [comment]
@ameshkov da, thats what i would say ^^

**Comment by ameshkov** [comment]
Just in case, here's everything I know about recursors and Go:
https://github.com/AdguardTeam/dnsproxy/issues/46

**Comment by ibksturm** [comment]
thats more than i know ^^ anyway i have go back to business

**Comment by ameshkov** [comment]
Experiments with CGO and unbound took A LOT of our time and the only thing that we understand now -- **we should avoid CGO** if we want to keep AGH truly cross-platform :(

Re-assigning this issue to v0.98, we'll make a second try now with a resolver written in pure go.

**Comment by lordraiden** [comment]
What is the status of this feature? How far are we or how dificult it will be?

**Comment by ameshkov** [comment]
It is quite difficult. We tried libunbound, but it's not the ideal solution for a number of reasons.

We started developing a pure Go implementation, but the problem is that it will take quite a lot of time to do it properly, and we have tons of other features we'd like to be implemented. 

The good news is that the author of CoreDNS seems to be going to implement a resolver in Go so there's a change that by the time we revisit this issue, there will be already something to work with.

**Comment by lordraiden** [comment]
> It is quite difficult. We tried libunbound, but it's not the ideal solution for a number of reasons.
> 
> We started developing a pure Go implementation, but the problem is that it will take quite a lot of time to do it properly, and we have tons of other features we'd like to be implemented.
> 
> The good news is that the author of CoreDNS seems to be going to implement a resolver in Go so there's a change that by the time we revisit this issue, there will be already something to work with.

Any news from coredns?

**Comment by ameshkov** [comment]
Nope:(

**Comment by hoshsadiq** [comment]
@ameshkov out of curiousity, Is there a reason [sdns resolver](https://github.com/semihalev/sdns/tree/master/middleware/resolver) isn't considered?

**Comment by ameshkov** [comment]
I am keeping an eye on it. I don't like that it depends on some SDNS server internal structures that aren't really needed for a resolver.

**Comment by ammnt** [comment]
@ameshkov, if this is implemented, will [DNS Zone Transfer-over-TLS/XFR-over-TLS](https://datatracker.ietf.org/doc/draft-ietf-dprive-xfr-over-tls/) be implemented as well?

**Comment by ameshkov** [comment]
Depends. The more I learn about DNS, the more I think we shouldn't implement a recursor ourselves, it's just too huge of a task. So if we find a recursive resolver library that fits, we'll use it and everything will depend on it.

**Comment by lordraiden** [comment]
> Depends. The more I learn about DNS, the more I think we shouldn't implement a recursor ourselves, it's just too huge of a task. So if we find a recursive resolver library that fits, we'll use it and everything will depend on it.

Core DNS is written in golang isn't the same than Adguard?

**Comment by ameshkov** [comment]
Core DNS is indeed written in golang, but it is not a recursor.

**Comment by lordraiden** [comment]
> Core DNS is indeed written in golang, but it is not a recursor.

There is this, but it looks dead
https://github.com/tenta-browser/tenta-dns

**Comment by ameshkov** [comment]
Yep, also their recursor implementation cannot be easily decoupled from the server's code, can't be used as a library.

**Comment by ammnt** [comment]
This function is still expected by me and, I think, only makes sense with the implementation of these two functions:
1. DNS Zone Transfer-over-TLS/XFR-over-TLS:
https://datatracker.ietf.org/doc/html/rfc9103
2. Root zone on loopback ("root hints"):
https://datatracker.ietf.org/doc/html/rfc7706
https://wiki.archlinux.org/title/Unbound#Root_hints

Lets consider it.

**Comment by erffy** [comment]
Any update on this?
